다형성으로 LoginService 인터페이스 구현

hamryt opened the pull request:
이슈번호 : #46

hamryt committed:
로그인 로직과 세션 로직 다른 클래스로 분리 및 다형성으로 LoginService 인터페이스 구현

hamryt committed:
Log4j2 설정

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/controller/LoginController.java:
```diff
@@ -23,7 +24,8 @@
 public class LoginController {
     
-    private final LoginService loginService;
+    private final LoginService<UserDTO> userLoginService;
```
이렇게 하면 결국 구체적인 타입에 의존하게되는데 로그인용 DTO를 통일시키고 제네릭을 제거하는게 나아보입니다~

hamryt committed:
loginService 리턴 타입 LoginDTO로 변경
다른 객체 타입에 의존하지 않기 위해 공통으로 사용하는 객체로 변경.
유저 타입 마다 다른 로그인 서비스 타입이 필요하므로 로그인 컨트롤러 메서드를 분리

hamryt committed:
Merge branch 'feature/43' into feature/46

hamryt committed:
merge develop

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/login/UserLoginServiceImpl.java:
```diff
@@ -14,24 +15,32 @@
-        return userDto;
+        
+        return LoginDTO.builder()
+            .id(userDTO.getId())
+            .email(userDTO.getEmail())
+            .name(userDTO.getName())
+            .phoneNumber(userDTO.getPhoneNumber())
+            .addressCode(userDTO.getAddressCode())
+            .addressDetail(userDTO.getAddressDetail())
+            .userType(userDTO.getUserType())
+            .build();
```
로그인 성공시 이렇게 많은 정보들을 클라이언트한테 왜 돌려줘야하는걸까요? 클라이언트에서 이 데이터를 이용해 어떤 동작을 시키려고 의도하신건지 궁금합니다~

hamryt committed:
loginFactory를 구현하여 로그인 컨트롤러에서 사용자별 login 서비스 클래스의 의존성을 분리

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/login/factory/LoginTypeFactory.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public LoginService createLoginService(UserType userType) {
+        LoginService loginService;
+        if (userType == UserType.USER){
```
switch-case로 해줘도 좋을 것 같아요~

hamryt committed:
가독성을 위한 switch문으로 팩토리 클래스 내 메서드 분기 방식 변경

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/login/factory/LoginTypeFactory.java:
```diff
@@ -16,12 +16,17 @@
+                loginService = userLoginService;
+                break;
+            case GYM:
+                loginService = gymLoginService;
```
굳이 이렇게 변수에 할당하기보다는 바로 리턴해줘도 될 것 같네요~

hamryt committed:
1. LoginTypeFactory 클래스 가독성 향상을 위한 코드스타일 변경
2. 해당 유저 타입에 관한 loginService가 없을 경우에 대한 예외 작성
